### PR TITLE
Authentification

### PR DESCRIPTION
**Pull request by vlf-stefan-drobic**
Hi @fedjo this is the second one of the divided PRs.

This one contains the authentification code, user registration, login etc. (offline/local)
I have an open issue for the gatekeeper integration, so I'm keeping that in mind, this PR doesn't include that, a future one will deal  with the integration.

### CODE REVIEW

File: app/core/security.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+from datetime import datetime, timedelta
+from typing import Any, Union
+from .config import settings
+from passlib.context import CryptContext
+
+import jwt
+
+ALGORITHM = "HS256"
+pwd_context = CryptContext(schemes=["argon2"])
```
Comment (fedjo) [inline app/core/security.py:9]: out of interest: why did you choose to use argon2 and not bcrypt?

File: app/crud/base.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+import traceback
+
+ModelType = TypeVar("ModelType", bound=Base)
+CreateSchemaType = TypeVar("CreateSchemaType", bound=BaseModel)
+UpdateSchemaType = TypeVar("UpdateSchemaType", bound=BaseModel)
+
+
+class CRUDBase(Generic[ModelType, CreateSchemaType, UpdateSchemaType]):
+    def __init__(self, model: Type[ModelType]):
```
Comment (fedjo) [inline app/crud/base.py:18]: why `Type[ModelType]` and not just `ModelType`?

**Review by fedjo — APPROVED** [review]

File: app/core/security.py (hunk 1 above)
Comment (vlf-stefan-drobic) [reply app/core/security.py:9]: I don't have one specific reason, it's mostly from reading online about it.

I did have a discussion with a colleague once, where we were discussing encryption technologies, and we had found out about argon2, so we used it in another project and it worked well, so we decided to keep using it for other projects as well I suppose.

Argon2 was already present in another requirements.txt file from another project, so I decided to reuse it. :)

File: app/crud/base.py (hunk 2 above)
Comment (vlf-stefan-drobic) [reply app/crud/base.py:18]: I'm not 100% sure anymore, I think it helps with typing when inheriting the class.
I did deploy the service with the "Type[]" part removed, and it worked the same, so I'm left to the previous conclusion

**Pull request merged**